File: ice0208/dp-1d/198.py
class Solution:
    def rob(self, nums: list[int]) -> int:
        if len(nums) == 1:
            return nums[0]
        if len(nums) == 2:
            return max(nums)

        nums[2] += nums[0]

        for i in range(3, len(nums)):
            nums[i] = nums[i] + max(nums[i - 2], nums[i - 3])

        return max(nums[-2:])


if __name__ == "__main__":
    print(Solution().rob([1, 2, 3, 1]))
    print(Solution().rob([2, 7, 9, 3, 1]))

File: moonpoison/dp_1demensional/198.cpp
#include<bits/stdc++.h>
using namespace std;
class Solution {
public:
    int rob(vector<int>& nums) 
    {
        vector<int> arr(401);
        if(nums.size()==0) return 0;
        if(nums.size()==1) return nums[0];
        if(nums.size()==2) return max(nums[0], nums[1]);
        arr[0]=nums[0]; arr[1]=nums[1]; arr[2]=nums[0]+nums[2];
        for(int i=3;i<nums.size();++i)
            arr[i]=max(arr[i-2], arr[i-3])+nums[i];
        return *max_element(arr.begin(), arr.end());
    }
};